### PR TITLE
Toric code: Add readout unfolding classes (#5)

### PR DESCRIPTION
**Pull request by dstrain115**

### CODE REVIEW

File: recirq/toric_code/readout_unfolding.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+Based on Nachman et al., "Unfolding quantum computer readout noise"
+npj Quantum Information 6 84 (2020) https://doi.org/10.1038/s41534-020-00309-7
+
+This is similar to using "readout matrix inversion" but more numerically stable.
+Note it does NOT take into account state preparation error, such as stray |1) population
+and pi pulse error.
+
+The readout matrix (R_ij in Nachman et al.) is the *transpose* of the arrangement in
+pyle.dataking.multiple_qubit.readout.correlated_readout_matrix.
```
Comment (mpharrigan) [inline recirq/toric_code/readout_unfolding.py:25]: I think this note is inappropriate for the ReCirq repository

File: recirq/toric_code/readout_unfolding.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ - Nachman: R_ij = Prob(measured=i | prepared=j) (beginning of "Methods" section)
+    → Columns sum to 1
+ - correlated_readout_matrix: data[row, col] = Prob(measured=col | prepared=row)
+    → Rows sum to 1
+
+We also must be consistent in bitstring ordering / endianness.
```
Comment (mpharrigan) [inline recirq/toric_code/readout_unfolding.py:31]: is this a statement of principles? what is the practical consequence

File: recirq/toric_code/toric_code_plotter.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
+# Copyright 2022 Google
```
Comment (mpharrigan) [inline recirq/toric_code/toric_code_plotter.py:1]: is this supposed to be a part of this PR? Do you need to merge or rebase?

File: recirq/toric_code/readout_unfolding.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+to use is not obvious (see Fig. 6), but in practice N=20 iterations works well.
+
+Nachman uses a uniform prior ("guess") probability distribution. This is the most "agnostic"
+choice possible. However, starting from the measured distribution works better empircally.
+"""
+
+import numpy as np
+
+
+class ReadoutUnfolder:
```
Comment (mpharrigan) [inline recirq/toric_code/readout_unfolding.py:43]: Not necessary to change, but I'd be loathe to not point out that this class could very well have been a free function

**Review by mpharrigan — APPROVED** [review]
lgtm, pending some minor fixes

**Review by mpharrigan — CHANGES_REQUESTED** [review]
actually I'm going to "request changes" until the merge issue is resolved so things don't accidentally get clobbered. will change back to approval once this is merged or rebased

File: recirq/toric_code/readout_unfolding.py (hunk 1 above)
Comment (dstrain115) [reply recirq/toric_code/readout_unfolding.py]: Removed.

File: recirq/toric_code/readout_unfolding.py (hunk 4 above)
Comment (dstrain115) [reply recirq/toric_code/readout_unfolding.py]: Done.

File: recirq/toric_code/readout_unfolding.py (hunk 2 above)
Comment (dstrain115) [reply recirq/toric_code/readout_unfolding.py:28]: I have no idea.  @kjsatz Can you answer?

**Review by mpharrigan — APPROVED** [review]
lgtm. @kjsatz can you speak to the one outstanding comment?

File: recirq/toric_code/readout_unfolding.py (hunk 2 above)
Comment (dstrain115) [reply recirq/toric_code/readout_unfolding.py]: Synced up.  The consistent ordering is needed for compatibility (between differing conventions in pyle and cirq).

**Pull request merged**